compliance/virtualmachine/relay.go

```diff
@@ -0,0 +1,236 @@
+
+	if err := r.vsockServer.Start(); err != nil {
+		return errors.Wrap(err, "starting vsock server")
+	}
+
+	go func() {
+		<-ctx.Done()
+		r.vsockServer.Stop()
+	}()
+
+	for {
+		conn, err := r.vsockServer.listener.Accept()
+		if err != nil {
+			if ctx.Err() != nil {
+				log.Info("Stopping virtual machine relay")
+				return ctx.Err()
+			}
+			log.Errorf("Error accepting connection: %v", err)
+			time.Sleep(time.Second) // Prevent a tight loop
+			continue
+		}
+
+		go func() {
+			if err := handleVsockConnection(ctx, conn, r.sensorClient); err != nil {
+				log.Errorf("Error handling vsock connection: %v", err)
+			}
+			if err := conn.Close(); err != nil {
+				log.Errorf("Failed to close connection: %v", err)
+			}
+
+		}()
+	}
+}
+
+func extractVsockCIDFromConnection(conn net.Conn) (uint32, error) {
+	remoteAddr, ok := conn.RemoteAddr().(*vsock.Addr)
+	if !ok {
+		return 0, fmt.Errorf("failed to extract remote address from vsock connection: unexpected type %T, value: %v",
+			conn.RemoteAddr(), conn.RemoteAddr())
+	}
+
+	return remoteAddr.ContextID, nil
+}
+
+func handleVsockConnection(ctx context.Context, conn net.Conn, sensorClient sensor.VirtualMachineIndexReportServiceClient) error {
+	metrics.VsockConnectionsAccepted.Inc()
+
+	log.Debugf("Handling vsock connection from %s", conn.RemoteAddr())
+
+	vsockCID, err := extractVsockCIDFromConnection(conn)
+	if err != nil {
+		return errors.Wrap(err, "extracting vsock CID")
+	}
+
+	maxSizeBytes := env.VirtualMachinesVsockConnMaxSizeKB.IntegerSetting() * 1024
+	timeout := 10 * time.Second
+	log.Debugf("Reading from connection (max bytes: %d, timeout: %s", maxSizeBytes, timeout)
+	data, err := readFromConn(conn, maxSizeBytes, timeout)
+	if err != nil {
+		return errors.Wrapf(err, "reading from connection (vsock CID: %d)", vsockCID)
+	}
+
+	log.Debugf("Parsing index report (vsock CID: %d)", vsockCID)
+	indexReport, err := parseIndexReport(data)
+	if err != nil {
+		return errors.Wrapf(err, "parsing index report data (vsock CID: %d)", vsockCID)
+	}
+	metrics.IndexReportsReceived.Inc()
+
+	// Fill the vsock context ID - at the moment the agent does not populate this field; if that changes, this can be
+	// replaced with a sanity check.
+	indexReport.VsockCid = strconv.Itoa(int(vsockCID))
+
+	if err = sendReportToSensor(ctx, indexReport, sensorClient); err != nil {
+		return errors.Wrapf(err, "sending report to sensor (vsock CID: %d)", vsockCID)
+	}
+
+	log.Debugf("Finished handling vsock connection from %s", conn.RemoteAddr())
+
+	return nil
+}
+
+func isRetryableGRPCError(err error) bool {
+	grpcErr, ok := status.FromError(err)
+	if !ok {
+		return false
+	}
+	code := grpcErr.Code()
+	switch code {
+	case codes.DeadlineExceeded:
+		return !errors.Is(err, context.Canceled)
+	case codes.Unavailable, codes.ResourceExhausted, codes.Internal:
+		return true
+	default:
+		return false
+	}
+}
+
+func parseIndexReport(data []byte) (*v1.IndexReport, error) {
+	report := &v1.IndexReport{}
```

Comment:
Actually it is v4.IndexReport in Scanner. What is the difference between this `v1.IndexReport` and the `v4.IndexReport` in stackrox?

Reply:
Ok, I see it. It may be still a bit confusing, but I think, it can stay, as we never had scanner `v1.IndexReport` (not even `v2` or `v3`).